ddtrace/tracer.py

```diff
@@ -462,9 +487,19 @@ def start_span(self, name, child_of=None, service=None, resource=None, span_type
                 self._runtime_worker.update_runtime_tags()
 
         self._hooks.emit(self.__class__.start_span, span)
-
         return span
 
+    def _on_span_finish(self, span):
+        if not self.enabled:
+            return  # nothing to do
+
+        if self.log.isEnabledFor(logging.DEBUG):
+            self.log.debug("writing span %r", span)
+
+        spans = self.processor.on_span_finish(span)
```

Comment:
I tried to implement this using a hook on `_on_span_finish` but we need to get a return value from the processor (or we'd have to fetch state from the processor in order to get the completed traces to write out) to pass to the writer.

@jd I think this was why I had added the `writer` attribute to `TraceProcessor` initially. I'm not sure if the spans/traces should need to come back to the `Tracer` since `Tracer` doesn't do anything additional with them other than forwarding them to a writer. The current approach here might also make having multiple writers/processors difficult. WDYT?

Reply:
Cannot this be implemented directly in `Span.finish`. At the moment, that seems to be the only caller of this hook. Besides, there seems to be a bit of an asymmetry in the way `on_span_start` and `_on_span_finish` are called: the former is called by the `Tracer` while the latter by the `Span`.

In fact, I would see this flushing logic better suited for one of the classes in `processor.py`, ideally the `Stitcher`, which might also be thought of as a `Flusher`

Reply:
> By this do you mean everything that is done currently in _on_span_finish?

Yep

Reply:
The `Tracer` should be the whole organizer of the data, so to me having the tracer passing the data to the processor, and then having them back (or not) to send to a writer, or whatever else is a good pick. It gives you flexibility.